### PR TITLE
added due_date for tasks

### PR DESCRIPTION
**Pull request by Kasvit**
**STATUS**
READY

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [ ] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [ ] PR doesn't have conflicts with main branch
- [ ] I've checked new feature as logged in and logged out user if needed
- [ ] "let!" is used in specs

### CODE REVIEW

File: app/helpers/account/tasks_helper.rb [hunk 1]
```diff
@@ -1,2 +1,13 @@
 module Account::TasksHelper
+  def task_style(section, duedate)
+    classes = ''
```
Comment (igorkasyanchuk) [inline app/helpers/account/tasks_helper.rb]: remove this variable it's not needed

File: app/helpers/account/tasks_helper.rb [hunk 2]
```diff
@@ -1,2 +1,13 @@
 module Account::TasksHelper
+  def task_style(section, duedate)
```
Comment (igorkasyanchuk) [inline app/helpers/account/tasks_helper.rb]: pass "task" to method (not section and duedate)

File: app/views/account/tasks/show.html.slim [hunk 3]
```diff
@@ -11,6 +11,9 @@
           = render('/account/tasks/watch_link', task: @task, project: @project)
     h3 Description
     = @task.description
+    - if !@task.due_date.nil?
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: if ...present?

File: app/views/account/tasks/_form.html.slim [hunk 4]
```diff
@@ -1,7 +1,8 @@
 = simple_form_for [:account, @project, @task] do |f|
   = f.input_field :section, as: :hidden, value: @task.section
   = f.input :title, placeholder: 'Title', wrapper: :bulma_input, required: true
-  = f.input :description, placeholder: 'Description', wrapper: :bulma_input
-  = f.input :files, input_html: { multiple: true }
+  = f.input :description, placeholder: 'Description', wrapper: :textarea
+  = f.input :due_date, as: :date, html5: true
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/_form.html.slim]: add some JS  date picker (not html5)

File: spec/factories/tasks.rb [hunk 5]
```diff
@@ -26,7 +27,7 @@
   factory :task, class: 'Task' do
     title { Faker::Lorem.sentence }
     description { Faker::Lorem.paragraph }
-
+    due_date { rand(1.month.ago..1.month.since)}
```
Comment (igorkasyanchuk) [inline spec/factories/tasks.rb]: also I don' want to have random due_dates
put nil

and add 2 traits with "expired" and "future" due dates

**Comment by Kasvit** [comment]
image
image

File: app/views/account/tasks/show.html.slim [hunk 6]
```diff
@@ -11,6 +11,9 @@
           = render('/account/tasks/watch_link', task: @task, project: @project)
     h3 Description
     = @task.description
+    - if @task.due_date.present?
+      h3 Due date
+      = @task.due_date.strftime("%d.%m.%Y")
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: use i18n and custom date format if needed

File: app/helpers/account/tasks_helper.rb [hunk 7]
```diff
@@ -1,2 +1,11 @@
 module Account::TasksHelper
+  def task_style(task)
+    if task.section?
+      'title is-4 is-italic'
+    elsif task.expired?
+      'has-text-danger'
+    else
+      ""
```
Comment (igorkasyanchuk) [inline app/helpers/account/tasks_helper.rb]: why this "" is needed?
remove

File: app/helpers/account/tasks_helper.rb [hunk 8]
```diff
@@ -1,2 +1,11 @@
 module Account::TasksHelper
+  def task_style(task)
```
Comment (romaslmd) [inline app/helpers/account/tasks_helper.rb]: is it style or class?

Comment (Kasvit) [reply app/helpers/account/tasks_helper.rb]: "style_for" is better?

Comment (romaslmd) [reply app/helpers/account/tasks_helper.rb]: usage in code is like `class: task_style(task)`
method name should reflect what's it does/returns
in your case it'd be more like `class: task_class(task)` or something like that

Comment (Kasvit) [reply app/helpers/account/tasks_helper.rb]: thanks

File: app/helpers/account/tasks_helper.rb [hunk 9]
```diff
@@ -1,2 +1,10 @@
 module Account::TasksHelper
+  def task_class(task)
+    if task.section?
+      'title is-4 is-italic'
+    elsif task.expired?
+      'has-text-danger'
+    else
```
Comment (igorkasyanchuk) [inline app/helpers/account/tasks_helper.rb:7]: why we need empty else? REMOVE!!!!!!

File: app/views/account/tasks/show.html.slim [hunk 10]
```diff
@@ -11,6 +11,9 @@
           = render('/account/tasks/watch_link', task: @task, project: @project)
     h3 Description
     = @task.description
+    - if @task.due_date.present?
+      h3 Due date
+      = I18n.localize(@task.due_date, format: :date_month_year_concise)
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: just use "l" helper (lower L)

File: app/views/account/projects/_show_tasks.html.slim [hunk 11]
```diff
@@ -5,7 +5,7 @@
         = link_to complete_account_project_task_path(project, task), class: 'add_to_complete', method: :patch, remote: true do
           = check_box_tag :complete
     .level-item
-      = link_to task.title, account_project_task_path(project, task), class: ('title is-4 is-italic' if task.section?)
+      = link_to task.title, account_project_task_path(project, task), class: task_class(task)
```
Comment (romaslmd) [inline app/views/account/projects/_show_tasks.html.slim]: probably `task_link_class` would make more sense..

**Comment by igorkasyanchuk** [comment]
And specs

**Comment by igorkasyanchuk** [comment]
Oh fix conflicts

File: app/models/task.rb [hunk 12]
```diff
@@ -47,6 +48,10 @@ def pending?
     !complete?
   end
 
+  def expired?
+    self.due_date && (self.due_date < Time.now && !complete?)
```
Comment (romaslmd) [inline app/models/task.rb]: nitpick `self.due_date && (self.due_date < Time.now && pending?)`
it's a lot easier to read/understand true-positive conditions
or more verbose `pending? && self.due_date && self.due_date < Time.now`

File: config/locales/en.yml [hunk 13]
```diff
@@ -31,3 +31,6 @@
 
 en:
   hello: 'Hello world'
+  time:
```
Comment (romaslmd) [inline config/locales/en.yml]: could you pls move this to a separate `formats.en.yml`

File: spec/factories/tasks.rb [hunk 14]
```diff
@@ -29,6 +30,7 @@
   factory :task, class: 'Task' do
     title { Faker::Lorem.sentence }
     description { Faker::Lorem.paragraph }
+    due_date { nil }
```
Comment (romaslmd) [inline spec/factories/tasks.rb]: it's nil by default
hm.. why nil? if i remember correctly faker has some date helpers..

File: spec/helpers/account/tasks_helper_spec.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  let!(:user)         { create(:user) }
+  let!(:workspace)    { create(:workspace, user: user) }
+  let!(:project)      { create(:project, workspace: workspace, users: [user]) }
+  let!(:section)      { create(:task, project: project, section: true) }
+  let!(:expired_task) { create(:task, :expired, project: project) }
+  let!(:task)         { create(:task, project: project) }
+
+  describe "#task_link_class" do
+    it "returns classes for section" do
+      expect(helper.task_link_class(section)).to eq('title is-4 is-italic')
```
Comment (romaslmd) [inline spec/helpers/account/tasks_helper_spec.rb]: try to test only on exact what you're adding with logically separate by contexts

```ruby
# no need to create so many records within a global scope.. they might not be needed in other helpers..

# describe something for the test
describe "#task_link_class" do
  # create all needed records within a scope for all contexts within a describe
  let!(:user) { create(:user) }
  let!(:workspace { create(:workspace, user: user) }
  let!(:project) { create(:project, workspace: workspace, users: [user]) }

  # separate on logical contexts
  context 'section' do
    # create records needed for specific context
    # this will be that data that depends on a use-case; e.g.: vary on a use-case
    let!(:task) { create(:task, project: project, section: true) }

    # declare a subject
    subject { helper.task_link_class(task) }
  
    # start testing..
    it { is_expected.to include('title') }
    it { is_expected.to include('is-4') }
    it { is_expected.to include('is-italic') }
  end

  # etc..
end
```

File: spec/factories/tasks.rb (hunk 14 above)
Comment (Kasvit) [reply spec/factories/tasks.rb]: its nil but in you can seed with traits :expired or :future 
due_date its an optional

Comment (igorkasyanchuk) [reply spec/factories/tasks.rb]: yes, nil is ok by default

File: Gemfile [hunk 16]
```diff
@@ -27,15 +27,16 @@ gem 'ransack'
...
 gem 'webpacker'
 gem 'react_on_rails'
 gem 'omniauth'
 gem 'omniauth-facebook'
 gem 'annotate'
 gem 'ranked-model'
 gem 'meta-tags'
-gem 'jquery-rails'
 gem "paranoia", '~> 2.2'
+# gem 'flatpickr_rails'
```
Comment (romaslmd) [inline Gemfile:39]: eh?

**Review by igorkasyanchuk — APPROVED** [review]

File: app/models/task.rb [hunk 17]
```diff
@@ -50,6 +51,10 @@ def pending?
     !complete?
   end
 
+  def expired?
+    self.due_date && (self.due_date < Time.now && pending?)
```
Comment (romaslmd) [inline app/models/task.rb]: no need to wrap in parentheses, condition works in one direction

File: config/locales/en.yml [hunk 18]
```diff
@@ -29,5 +29,4 @@
 # To learn more, please read the Rails Internationalization guide
 # available at http://guides.rubyonrails.org/i18n.html.
 
-en:
-  hello: 'Hello world'
+en:
```
Comment (romaslmd) [inline config/locales/en.yml:32]: unnecessary change

File: spec/factories/tasks.rb [hunk 19]
```diff
@@ -45,5 +46,12 @@
       end
     end
 
+    trait :expired do 
+      due_date { Faker::Date.backward(30) }
+    end
+
+    trait :future do 
+      due_date { Faker::Date.forward(30) }
```
Comment (romaslmd) [inline spec/factories/tasks.rb:54]: nice 👍

File: Gemfile (hunk 16 above)
Comment (Kasvit) [reply Gemfile:39]: its for next task "flatpickr" datepicker

File: app/models/task.rb (hunk 17 above)
Comment (Kasvit) [reply app/models/task.rb]: I know it but it's more readable for me like "true && true" than "true && true && true"

File: spec/helpers/account/tasks_helper_spec.rb [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+  let!(:project)      { create(:project, workspace: workspace, users: [user]) }
+
+  context 'section' do
+    let!(:task) { create(:task, project: project, section: true) }
+
+    subject { helper.task_link_class(task) }
+  
+    it { is_expected.to include('title') }
+    it { is_expected.to include('is-4') }
+    it { is_expected.to include('is-italic') }
```
Comment (romaslmd) [inline spec/helpers/account/tasks_helper_spec.rb:15]: `its(:split) { is_expected.to contain_exactly('title', 'is-4', 'is-italic') }`
